src/paymentservice/src/main/java/ca/bc/gov/open/pssg/rsbc/dps/paymentservice/BamboraClientImpl.java

```diff
@@ -0,0 +1,184 @@
+		return hostedPaymentURL;
+	}
+
+	public String getMerchantId() {
+		return merchantId;
+	}
+
+	public URL getHostedPaymentURL() {
+		return hostedPaymentURL;
+	}
+
+	public int getMinutesToExpire() {
+		return minutesToExpire;
+	}
+	
+	public String getCorrelationId() {
+		return correlationId;
+	}
+
+	// constructor
+	public BamboraClientImpl(URL hostedPaymentURL, String merchantId, String hashKey, int minutesToExpire) throws PaymentServiceException {
+		this.hostedPaymentURL = hostedPaymentURL;
+		this.merchantId = merchantId;
+		this.hashKey = hashKey;
+		this.minutesToExpire = minutesToExpire;
+	}
+
+	/**
+	 * 
+	 * Calculate Single Payment URL 
+	 * 
+	 * @param correlationId
+	 * @param transType
+	 * @param invoiceNumber
+	 * @param totalItemsAmount
+	 * @param totalGST
+	 * @param totalPST
+	 * @param approvedPage
+	 * @param declinedPage
+	 * @param errorPage
+	 * @param ref1
+	 * @param ref2
+	 * @param ref3
+	 * @return
+	 * @throws PaymentServiceException
+	 */
+	@Override
+	public URL calculateSinglePaymentURL(String correlationId, BamboraTransType transType, String invoiceNumber, double totalItemsAmount, double totalGST, 
+			double totalPST, String approvedPage, String declinedPage, String errorPage, String ref1, String ref2, String ref3)
+			throws PaymentServiceException {
+		
+		// add correlationId if non provided in request. 
+		if (StringUtils.isEmpty(correlationId)) {
+			this.correlationId = generateUniqueCorrelationId();
+	    }
+		
+		String redirect = null;
+		
+		try {
+	
+			// Add the optional parameters if available.
+			String errorParam = (errorPage != null) ? "&" + PaymentServiceConstants.BAMBORA_PARAM_ERROR_PAGE + "=" + errorPage
+					: "";
+			String declinedParam = (approvedPage != null)
+					? "&" + PaymentServiceConstants.BAMBORA_PARAM_APPV_PAGE + "=" + approvedPage
+					: "";
+			String approvedParam = (declinedPage != null)
+					? "&" + PaymentServiceConstants.BAMBORA_PARAM_DECL_PAGE + "=" + declinedPage
+					: "";
+			String endUserIdParam = (ref1 != null) ? "&" + PaymentServiceConstants.BAMBORA_PARAM_REF1 + "=" + ref1 : "";
+			String statementParam = (ref2 != null) ? "&" + PaymentServiceConstants.BAMBORA_PARAM_REF2 + "=" + ref2 : "";
+			String echoParam = (ref3 != null) ? "&" + PaymentServiceConstants.BAMBORA_PARAM_REF3 + "=" + ref3 : "";
+	
+			// Create the base url
+			String paramString = PaymentServiceConstants.BAMBORA_PARAM_MERCHANT_ID + "=" + this.merchantId + "&"
+					+ PaymentServiceConstants.BAMBORA_PARAM_TRANS_TYPE + "=" + transType.toString() + "&"
+					+ PaymentServiceConstants.BAMBORA_PARAM_TRANS_ORDER_NUMBER + "=" + invoiceNumber +
+					errorParam + approvedParam + declinedParam + endUserIdParam + statementParam + echoParam +
+						"&" + PaymentServiceConstants.BAMBORA_PARAM_TRANS_AMOUNT + "="
+					+ String.format("%1$,.2f", totalItemsAmount + totalPST + totalGST);
+	
+			// Replace spaces with escaped equivalent
+			paramString = paramString.replace(" ", "%20");
```

Comment:
Should we build the paramString as an array and do a join ?

Reply:
Done the best I could. I also attempt to keep this section fairly clean as assembling many parameters into one call can be confusing for another dev to look at later.